### ISSUE TITLE
tf.estimator generator_input_fn multi thread bug 

### ISSUE THREAD

**Issue by Scitator** [tensorflow/tensorflow#10597]
I would like to use python generators as input pipeline for tf.estimator's. Finally I found generator_input_fn (announced [here](https://youtu.be/5DknTFbcGVM?t=12m30s)).

Nevertheless, when I start first experiments, I found I bug with learning curve. After some experiments I found the purpose of it - all blows up if you set `num_threads` > 1 in generator_input_fn.

For example, here is loss plot with python generator and `num_threads=1`:
![tf_py_generator_1_thread](https://user-images.githubusercontent.com/7606451/26978620-d58faec2-4d34-11e7-8f8b-14e4e50f6844.png)

And with `num_threads=2`:
![tf_py_generator_2_threads](https://user-images.githubusercontent.com/7606451/26978619-d58a9e3c-4d34-11e7-8f1b-8bab67a957aa.png)

**But** if I use `tf.estimator.inputs.numpy_input_fn` with 4 thread all work pretty well:
![tf_numpy_generator_4_threads](https://user-images.githubusercontent.com/7606451/26978621-d591c11c-4d34-11e7-85ce-2340bd2ed01c.png)
Except the fact, that I cannot save all my data in one numpy array (GBs of data).

Any suggestions why so? Or I need just wait for TF 1.2 with working `tf.estimators.inputs.generator_input_fn`?

Jupyter notebook with my experiments [here](https://gist.github.com/Scitator/184c8d676f36a9b7c04fb504d9088590).

**Comment by girving** [comment]
Unfortunately, I think we would need more information to have a sense where the problem is.  Learning curves that explode are often subtle machine learning issues, not bugs in TensorFlow.  It's entirely possible that it is a bug, though.  Could you do some more debugging to try to narrow down the cause?

**Comment by Scitator** [comment]
First of all, I always use grad clipping to prevent such problem, and come on, it's typical conv net for mnist :)
And secondly, I already done some debugging before reporting an issue:
Different combinations of `queue_capacity`, `num_epochs`, `shuffle` have no affect.
`sparse_softmax_cross_entropy_with_logits` also works correct (I tried with just `softmax_cross_entropy_with_logits` - still have such issue).
Python generator work pretty well, without any exceptions, *but* it still strange, that loss always explode at the end of first iteration over dataset.

That's why I think, that something go wrong with multi-threading here. Nevertheless, you can always check notebook and run it.

**Comment by girving** [comment]
@martinwicke Want to investigate?  I am unfamiliar with possible `tf.learn` threading issues.

**Comment by martinwicke** [comment]
This is a bug in generator_input_fn -- I just looked and if I read it correctly, it makes a generator per thread (each thread calls the generator function passed to generator_input_fn). 

If you'd like to fix that, we'd welcome a fix (although it looks non-trivial). Even a doc fix would be welcome.

The alternative is to write your generator to not use global state (in your case, `data_src.next_batch(1)` -- give each generator a separate copy of data_src).

**Comment by Scitator** [comment]
@martinwicke Thanks, now I understand. **And it's not a bug, it's a feature :)**
Python generator cannot be used from several threads, so the best way to give each thread it own one.
But now we have an issue, when the data from threads is the same. For me, I solve it through adding a thread number parameter in `generator_input_fn` and split data correctly in it. [link](https://github.com/Scitator/rstools/blob/master/rstools/tf/data_iterator.py). 
As I understand TF source code, it uses different seeds to prevent such same data yielding. Although, it only set it for python random, strange a bit. Looks like all this done for better code reusing, or it really works correctly....need to test.

Nevertheless, I cannot understand such loss explosion. But still, thanks for help.

**Comment by martinwicke** [comment]
Yes, you have to take care of data splitting yourself. Shuffling should work fine -- unless you set the random seed in each of the threads.

I don't know what the loss is doing, I can imagine that it does all kinds of terrible things, for instance it may return all zeros, and only because you use gradient clipping do you even get a non-inf loss. I don't know. I'm pretty sure you didn't see valid data after the first epoch.

**Comment by bhack** [comment]
I think that we could close this as we are not going to have features evolution for  `tf.estimator` right?

**Comment by sushreebarsa** [comment]
@Scitator Estimators are not recommended for new code , could you please refer this [link](https://www.tensorflow.org/guide/estimator) and let us know if we can move this issue to closed status?
Thanks!
